intro-graceful.Rmd

```diff
@@ -0,0 +1,45 @@
+# Graceful HTTP R packages
+
+Based on the previous chapter, your package interacting with a web resource has a dependency on `{curl}`, `{httr}` or `{crul}`. You have hopefully read the docs of the dependency you chose, including, in the case of httr and crul, the vignette about best practice for HTTP packages. Now, in this chapter we want to give more tips aimed at making your HTTP R package graceful, part of which you'll learn more about in this very book!
+
+**Why** write a *graceful* HTTP R package? First of all, graceful is a nice adjective. 💃🕺Then, graceful is the adjective used in [CRAN repository policy](https://cran.r-project.org/web/packages/policies.html) *"Packages which use Internet resources should fail gracefully with an informative message if the resource is not available or has changed (and not give a check warning nor error)."* Therefore, let's review how to make your R package graceful from this day forward, in success and in failure.
+
+## Choose the HTTP resource wisely
+
+First of all, your life and the life of your package's users will be easier if the web service you're wrapping is well maintained and well documented. When you have a choice, try not to rely on a fragile web service. Moreover, if you can, try to communicate with the API providers (telling them about your package; reporting feature requests and bug reports in their preferred way).
+
+## User-facing grace (how your package actually works)
+
+0.  If you can, do not request the API every time the user asks for something but cache data instead. No API call, no API call failure! 😉 To remember answers within a session check out [memoise](<https://github.com/r-lib/memoise>). To remember answers across sessions, see [webmiddens](https://docs.ropensci.org/webmiddens/), and approaches presented in the R-hub blog post ["Persistent config and data for R packages"](<https://blog.r-hub.io/2020/03/12/user-preferences/>)
```

Comment:
probably mention caveat here that package maintainers should be careful about caching responses. caching behavior should be well documented for users, and there should probably be an expiration time for caches that's based on how often data is updated on the remote service.